internal/pkg/mount/v3/manager.go

```diff
@@ -0,0 +1,227 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package mount
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"golang.org/x/sys/unix"
+
+	"github.com/siderolabs/talos/internal/pkg/xfs"
+	"github.com/siderolabs/talos/internal/pkg/xfs/fsopen"
+)
+
+// Manager is the filesystem manager for mounting and unmounting filesystems.
+type Manager struct {
+	fs xfs.FS
+
+	target  string
+	printer func(string, ...any)
+
+	selinuxLabel          string
+	shared                bool
+	skipIfMounted         bool
+	keepOpen              bool
+	mountattr             int
+	extraDirs             []string
+	extraUnmountCallbacks []func(m *Manager)
+
+	point *Point
+}
+
+// NewManager creates a new Manager with the given options.
+func NewManager(opts ...ManagerOption) *Manager {
+	m := &Manager{}
+
+	for _, opt := range opts {
+		opt.set(m)
+	}
+
+	return m
+}
+
+// Mount mounts a filesystem with the given options.
+func (m *Manager) Mount() (*Point, error) {
+	printer := discard
+	if m.printer != nil {
+		printer = m.printer
+	}
+
+	for _, dir := range m.extraDirs {
+		printer("creating directory tree %q", dir)
+
+		if !filepath.IsAbs(dir) {
+			return nil, fmt.Errorf("dir %q is not absolute", dir)
+		}
+
+		if err := os.MkdirAll(dir, 0o755); err != nil {
+			return nil, fmt.Errorf("error creating mount point directory %q: %w", dir, err)
+		}
+	}
+
+	root := &xfs.UnixRoot{FS: m.fs}
+
+	if err := root.OpenFS(); err != nil {
+		return nil, fmt.Errorf("openfs failed: %w", err)
+	}
+
+	m.point = &Point{
+		root:         root,
+		keepOpen:     m.keepOpen,
+		target:       m.target,
+		selinuxLabel: m.selinuxLabel,
+	}
+
+	opts := Options{
+		Printer:         printer,
+		Shared:          m.shared,
+		SkipIfMounted:   m.skipIfMounted,
+		MountAttributes: m.mountattr,
+	}
+
+	printer("mkdirAll %q", m.target)
+
+	if err := os.MkdirAll(m.target, 0o755); err != nil {
```

Comment:
not for this PR, but I wish we dropped this and did more explicit mkdir based on specific settings